### PR TITLE
Create ZeroPool.md

### PR DESCRIPTION
**Pull request by snjax**
# Grant Application Checklist

- [x] The [application-template.md](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( "project_name.md") and updated.
- [x] A BTC or Ethereum address for the payment of the milestones is provided inside the application.  
- [x] The software of the project will be released under the Apache license version 2.0 as specified in the [terms and conditions](https://gist.github.com/Noc2/75bc58e8ce9b5d419ff883b0cf2b8c19).
- [x] The total funding amount of the project is below USD $30k for initial grant applications and $100k for follow-up grants at the time of submission.
- [x] The initial PR contains only one commit (squash if needed before submitting your PR).
- [x] The grant will only be announced once we successfully [delivered](https://github.com/w3f/Grant-Milestone-Delivery) the first milestone.

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the applications. A few questions from my side (also based on some internal feedback): 

- You still need to sign the Terms and conditions.
- Could you try to leverage the [open grants template for the application](https://github.com/w3f/Open-Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt). Specifically regarding the milestones.
- Has ZeroPool received any funding/further support from the Ethereum Foundation to develop their research?
- The application is pretty ethereum focused. "ZeroPool supports ERC20 and ERC721". Are there any adjustments planned for Polkadot?
- Has any of the ZeroPool (ethereum) code been audited so far?

**Comment by burdges** [comment]
What is the ZeroPool design?  Zcash-like?  Spends reference UTXOs via a Merkle tree.  And nullifiers accumulate?  That'd be fine.  

Any cryptography could run in WASM on substrate of course, but..

We'll eventually provide native code support for BLS12-377, BW6, and their JubJubs via Arkworks.  If you're interested in helping there then take a look at https://github.com/arkworks-rs/curves/issues/17 

We'll probably never provide any native code support for any non-arkworks curves implementations besides our existing secp256k1 and ed25519/ristretto code.

**Comment by snjax** [comment]
@burdges,

>What is the ZeroPool design? Zcash-like? Spends reference UTXOs via a Merkle tree. And nullifiers accumulate on-chain? That'd be fine.

It is like ZCash, but there are some differences for better support of account-based chains. For example, with such privacy as ZCash we publish only one nullifier per transaction. So, storage space is used more sparingly.

[My talk on ZKSummit about account-based privacy](https://www.youtube.com/watch?v=f885LTdgJVs)
[Account-based private transaction engine](https://ethereum-magicians.org/t/account-based-private-transaction-engine/4833)

> Any cryptography could run in WASM on substrate of course, but.

Our current plans are building protocol at the most widespread curve and proving system (currently it is groth16 and bn256). With the current codebase, we can deploy the working solution faster and safer.

After that, we are going to switch our research to more novel protocols. 1-level recursion on BLS12-377 and BW6 is promising. But novel protocols could be better.

**Comment by burdges** [comment]
We've no timeframe for doing anything native, so yes if your code works well in WASM then that'll definitely be faster, and sure bn254 might run faster in WASM than BLS12-377 or whatever.  I mentioned https://github.com/arkworks-rs/curves/issues/17 largely because it might fit your team's interest.

**Comment by burdges** [comment]
Apologies for the delay.. Yes, we should fund this project @Noc2 

@snjax  As you're doing zk, you could do you own off-chain storage likely just block contents aka UTXOs, but also you could use Poseidon integrated with substrate's storage in your own substrate fork, and then only provide SNARKs that get checked in the runtime (eventually on polkadot).  I'd think either of the top contributors to https://github.com/paritytech/trie/graphs/contributors could give advise on this if you need anything like that.  We're quite slow about upstreaming improvements to substrate's storage architecture, but we'd love to see exploration of such options.

**Comment by burdges** [comment]
I'm confused by one point, what do you mean by "optimistic rollup"?

**Comment by snjax** [comment]
@burdges,

>what do you mean by "optimistic rollup"?

Our current beta implementation on Ethereum is working on L2 (custom optimistic rollup). In optimistic rollup, the relayer just publishes transactions and result of VM execution, but do not execute them. If the result is wrong, anybody can challenge the relayer by executing the transaction and stop the rollup. So, an expensive zkSNARK verifier is executed only if the relayer publishes damaged data.

>if I understand, users supply SNARKs for UTXOs but you'll support accounts too?

Yes, the account is a special type of UTXO, containing spend interval (in the simplest case it is maximal UTXO index spendable by the user). So, when we perform the transfer, it is enough only to publish only one nullifier for an account and increase the spent interval for the account.

>you could use Poseidon integrated with substrate's storage in your own substrate fork, and then only provide SNARKs that get checked in the runtime (eventually on polkadot)

Sure, we are going to compute Poseidon only off-chain and verify Merkle root update with zkSNARKs. Poseidon is very expensive to compute it onchain.

**Comment by Noc2** [comment]
@snjax would it be possible to reduce the price. This way it will be easier to get the necessary approval.

**Comment by snjax** [comment]
> 
> 
> @snjax would it be possible to reduce the price. This way it will be easier to get the necessary approval.

@Noc2, now we can reduce the total amount by removing the last milestone. Also, we start thinking about publishing our SDK, so other developers could implement their own dApps, based on our protocol.

**Comment by Noc2** [comment]
@snjax Thanks for updating the application. Removing the last milestone makes a lot of sense. Especially since we can always sign a follow-up grant for it, but I was more referring to the general costs of the proposal. For example Milestone 2 and 3 both currently have the same estimated duration and FTE and milestone 2 costs 40k instead of 30k. It would be appreciated if you could reduce the price of the deliveries a little bit more and would help to get the necessary approval from the rest of the committee.

**Comment by Noc2** [comment]
@snjax Any updates on this?

**Comment by snjax** [comment]
@Noc2 
We very much appreciate the cooperation with you and the work we have done. I think we are able to consider 63k dai as a total grant amount for current milestones.

The costs are different because each step requires different people.

**Comment by Noc2** [comment]
@snjax Thanks for the quick response. Sounds good to me. Could you update the application?

**Comment by snjax** [comment]
@Noc2  done

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I’m happy to go ahead with it and I will share the application with the rest of the committee.

**Review by mmagician — APPROVED** [review]
The previous delivery was good, I'm also happy to proceed with a follow-up

**Review by BenWhiteJam — APPROVED** [review]

**Comment by alxs** [comment]
Hey @snjax, care to share a quick status update on the project?

**Comment by snjax** [comment]
Hi, @alxs. Now we are focused on the core part of development (cryptography, zkSNARKs, and documentation). We decided to go step by step. Now we prepared zkSNARKs for the audit.
So, we could be sure, we reduce the security risks as more as we can before the release. And further development is going to be more smooth and more effective.

**Comment by alxs** [comment]
Thanks for the update, sounds like a wise decision. If you expect M2 to still take a while, I recommend you submit an amendment and modify its duration accordingly.

**Comment by semuelle** [comment]
It sounds like it might still take some time before you can submit a milestone, @snjax. Could you submit a PR to your grant application with an updated timeline?

File: applications/ZeroPool.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+## **Development Roadmap**
+
+* **Total Estimated Duration:** 4 months
+* **Full-time equivalent (FTE):**  3 FTE
+* **Total Costs:** 63000 DAI
+
+### Milestone 2 — Implement zkSNARK circuit and cryptography for private transaction
+
+* **Estimated Duration:** 2 months
```
Comment (alxs) [inline applications/ZeroPool.md:100]: Please @snjax, submit an amendment. The duration should be updated to 5 months (which would be now) + however long you think you will still need.

**Comment by alxs** [comment]
@snjax this is a last request to please submit an amendment. Otherwise, we'll terminate the grant in the coming days.

**Comment by alxs** [comment]
@snjax a reminder to submit milestone 2 - please follow the process above.

**Comment by snjax** [comment]
@alxs Sorry for the late reply, it was LisCon and big backlog. I submitted the milestone 2 here https://github.com/w3f/Grant-Milestone-Delivery/pull/316

**Pull request merged**